File: Έγγραφα SEM4105/Λύσεις ασκήσεων/09. Ένατο εργαστήριο/6/hermite.cpp
#include <iostream>
#include <cmath>
#include <array>


double factorial(int n)
{
    return (n==0 ? 1.0 : factorial(n-1)*n);
}



double hermite(int n, double x)
{
    double r;
    
    switch (n) {
    case 0:
	r = 1.0;
	break;
    case 1:
	r = 2.0 * x;
	break;
    case 2:
	r = 4.0 * x * x - 2.0;
	break;
    case 3:
	r = x * (8.0 * x * x - 12.0);
	break;
    case 4:
	r = x * x * (16.0 * x * x - 48.0) + 12.0;
	break;
    default:
	std::cerr << "not implemented\n";
	break;
    }

    return r;
}
      

double weight(int n, double x)
{
    double constexpr pi{3.14159265358979323846};
    
    return std::pow(2.0,n-1) * factorial(n) * std::sqrt(pi)
	/ std::pow(n * hermite(n-1,x), 2);
  
}


double f(double x)
{
    return x*x;
}



int main()
{
    std::szie_t constexpr n{4};
    double constexpr pi{3.14159265358979323846};
    
    std::array<double,n> x;
    x[0] = std::sqrt((3.0-std::sqrt(6.0)) / 2.0);
    x[1] = std::sqrt((3.0+std::sqrt(6.0)) / 2.0);
    x[2] = -x[0];
    x[3] = -x[1];

    double s{0.0};
    for (std::array<double,n>::size_type i{0}; i < n; ++i) {
	double const w = weight(n, x[i]);
	double const fval = f(x[i]);
	s += w * fval;
    }

    std::cout.precision(15);

    std::cout << "the integral is " << s << '\n'
	      << "the correct value is "<< std::sqrt(pi) / 2.0 << '\n';
    std::cout << "exactly the same (why?)\n";

}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/09. Ένατο εργαστήριο/6/hermite.f
      PROGRAM GAUHER
      INTEGER N, I
      PARAMETER (N=4)
      DOUBLE PRECISION X(N), W, FVAL, SUM
      DOUBLE PRECISION PI
      PARAMETER (PI=3.14159265358979323846D0)
      EXTERNAL F, WEIGHT
      DOUBLE PRECISION F, WEIGHT

      X(1) = SQRT ( (3.0D0-SQRT(6.0D0)) / 2.0D0 )
      X(2) = -X(1)
      X(3) = SQRT ( (3.0D0+SQRT(6.0D0)) / 2.0D0 )
      X(4) = -X(3) 
      
      SUM = 0.0D0
      DO I=1,N
        W = WEIGHT(N, X(I))
        FVAL = F(X(I))
        SUM = SUM + W * FVAL
      ENDDO

      PRINT *, "The integral is", SUM
      PRINT *, "The correct value is", SQRT(PI) / 2.0D0
      PRINT *, "Exactly the same (Why?)"
      END

      FUNCTION F(X)
      DOUBLE PRECISION X,F
      F = X*X
      END

      FUNCTION PARAG(N)
      INTEGER N, PARAG
      INTEGER TEMP, I

      TEMP = 1
      DO I=2,N
         TEMP = TEMP * I
      ENDDO
      PARAG = TEMP
      END

      FUNCTION WEIGHT(N, X)
      INTEGER N
      DOUBLE PRECISION X, WEIGHT
      EXTERNAL HERMIT
      DOUBLE PRECISION HERMIT
      DOUBLE PRECISION PI
      PARAMETER (PI=3.14159265358979323846D0)
      EXTERNAL PARAG
      INTEGER PARAG
  
      WEIGHT = 2.0D0**(N-1) * PARAG(N) * SQRT(PI) /
     &     (N * HERMIT(N-1,X))**2

      END

      FUNCTION HERMIT(N, X)
      INTEGER N
      DOUBLE PRECISION X, HERMIT

      DOUBLE PRECISION RES
      
      IF (N.EQ.0) RES = 1.0D0
      IF (N.EQ.1) RES = 2.0D0 * X
      IF (N.EQ.2) RES = 4.0D0 * X * X - 2.0D0
      IF (N.EQ.3) RES = X * (8.0D0 * X * X - 12.0D0)
      IF (N.EQ.4) RES = X * X * (16.0D0 * X * X - 48.0D0) + 12.0D0
      IF (N.GE.5) STOP "NOT IMPLEMENTED"

      HERMIT = RES

      END
